File: server/internal/domain/search/contract.go
package search

import (
	"context"
	"encoding/json"

	"github.com/shopspring/decimal"

	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/strategy"
)

type SearchSpace struct {
	StrategyIDs   []string                    `json:"strategy_ids"`
	ParameterGrid map[string]map[string][]any `json:"parameter_grid"`
	Cardinality   [2]int                      `json:"cardinality"`
	Policies      []string                    `json:"policies"`
	WeightOptions []decimal.Decimal           `json:"weight_options"`
}

type ScoredCandidate struct {
	Candidate strategy.CandidateStrategy
	Score     decimal.Decimal
}

type HashSet interface {
	Has(string) bool
	Add(string)
}

type SearchHistory struct {
	TestedHashes      HashSet
	TopK              []ScoredCandidate
	BestScore         *decimal.Decimal
	NonImprovingCount int
}

type History = SearchHistory

type CandidateGenerator interface {
	GeneratorID() string
	GeneratorVersion() string
	Generate(context.Context, SearchSpace, int, *int64, SearchHistory) ([]strategy.CandidateStrategy, error)
}

type StopConditions struct {
	MaxCandidates   *int             `json:"max_candidates,omitempty"`
	MaxDurationSec  *int             `json:"max_duration_sec,omitempty"`
	MaxNonImproving *int             `json:"max_non_improving,omitempty"`
	MaxFailureRate  *decimal.Decimal `json:"max_failure_rate,omitempty"`
}

type SearchRun struct {
	GeneratorID      string          `json:"generator_id"`
	GeneratorVersion string          `json:"generator_version"`
	StopConditions   StopConditions  `json:"stop_conditions"`
	Seed             *int64          `json:"seed,omitempty"`
	ExecutionConfig  json.RawMessage `json:"execution_config"`
}

File: py_backend/app/domain/search/contract.py
"""Search contracts (float64).

Mirrors `server/internal/domain/search/contract.go`.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Protocol

from ..common import Decimal
from ..strategy import CandidateStrategy


@dataclass
class SearchSpace:
    strategy_ids: list[str] = field(default_factory=list)
    parameter_grid: dict[str, dict[str, list[Any]]] = field(default_factory=dict)
    cardinality: tuple[int, int] = (0, 0)
    policies: list[str] = field(default_factory=list)
    weight_options: list[Decimal] = field(default_factory=list)


@dataclass
class ScoredCandidate:
    candidate: CandidateStrategy
    score: Decimal


class HashSet(Protocol):
    def has(self, key: str) -> bool: ...

    def add(self, key: str) -> None: ...


@dataclass
class SearchHistory:
    tested_hashes: HashSet | None = None
    top_k: list[ScoredCandidate] = field(default_factory=list)
    best_score: Decimal | None = None
    non_improving_count: int = 0


History = SearchHistory


class CandidateGenerator(Protocol):
    def generator_id(self) -> str: ...

    def generator_version(self) -> str: ...

    def generate(
        self,
        space: SearchSpace,
        batch: int,
        seed: int | None,
        history: SearchHistory,
    ) -> list[CandidateStrategy]: ...


@dataclass
class StopConditions:
    max_candidates: int | None = None
    max_duration_sec: int | None = None
    max_non_improving: int | None = None
    max_failure_rate: Decimal | None = None


@dataclass
class SearchRun:
    generator_id: str
    generator_version: str
    stop_conditions: StopConditions = field(default_factory=StopConditions)
    seed: int | None = None
    execution_config: Any | None = None
